src/hugr/views/sibling.rs

```diff
@@ -0,0 +1,562 @@
+/// [`super::SiblingGraph`],  not all nodes of the sibling graph must be
+/// included. A convex subgraph is always an induced subgraph, i.e. it is defined
+/// by a set of nodes and all edges between them.
+
+/// The incoming boundary (resp. outgoing boundary) is given by the input (resp.
+/// output) ports of the subgraph that are linked to nodes outside of the subgraph.
+/// The signature of the subgraph is then given by the types of the incoming
+/// and outgoing boundary edges. Given a replacement with the same signature,
+/// a [`SimpleReplacement`] can be constructed to rewrite the subgraph with the
+/// replacement.
+///
+/// The ordering of the nodes in the subgraph is irrelevant to define the convex
+/// subgraph, but it determines the ordering of the boundary signature.
+#[derive(Clone, Debug)]
+pub struct SiblingSubgraph<'g, Base> {
+    /// The underlying Hugr.
+    base: &'g Base,
+    /// The nodes of the induced subgraph.
+    nodes: Vec<Node>,
+}
+
+impl<'g, Base: HugrInternals> SiblingSubgraph<'g, Base> {
+    /// A sibling subgraph from a [`crate::ops::OpTag::DataflowParent`]-rooted HUGR.
+    ///
+    /// The subgraph is given by the nodes between the input and output
+    /// children nodes of the parent node. If you wish to create a subgraph
+    /// from another root, wrap the `region` argument in a [`super::SiblingGraph`].
+    ///
+    /// This will return an [`InvalidSubgraph::EmptySubgraph`] error if the
+    /// subgraph is empty.
+    pub fn from_dataflow_graph<Root>(dfg_graph: &'g Base) -> Result<Self, InvalidSubgraph>
+    where
+        Base: HugrView<RootHandle = Root>,
+        Root: ContainerHandle<ChildrenHandle = DataflowOpID>,
+    {
+        let parent = dfg_graph.root();
+        let nodes = dfg_graph.children(parent).skip(2).collect_vec();
+        if nodes.is_empty() {
+            Err(InvalidSubgraph::EmptySubgraph)
+        } else {
+            Ok(Self {
+                base: dfg_graph,
+                nodes,
+            })
+        }
+    }
+
+    /// Create a new sibling subgraph from some boundary edges.
+    ///
+    /// Any sibling subgraph can be defined using two sets of boundary edges
+    /// $B_I$ and $B_O$, the incoming and outgoing boundary edges respectively.
+    /// Intuitively, the sibling subgraph is all the edges and nodes between
+    /// an edge of $B_I$ and an edge of $B_O$.
+    ///
+    /// The `incoming` and `outgoing` arguments give $B_I$ and $B_O$ respectively.
+    /// They can be either source or target ports. We currently assume that if
+    /// the source port of an outgoing boundary edge is linked to multiple
+    /// target ports, then all edges from the same source port are outgoing
+    /// boundary edges.
+    ///
+    /// More formally, the sibling subgraph defined by $B_I$ and $B_O$ is the
+    /// graph given by the connected components of the graph with edges
+    /// $E\B_I\B_O$ which contain at least one node that is either
```

Comment:
nit: note we haven't defined `$E$` here! I quite liked the "connected components of the graph G'=(V, E \ B_I \ B_O)` too as it's then clearer that "which contain at least one node..." refers to the connected components rather than the edges. (But it's not really ambiguous as edges don't contain nodes so ok.)